### PR TITLE
Add Lutra script

### PR DESCRIPTION
**Pull request by chrisib**
Adds a new syncable LFO script, similar to e.g. Harmonic LFO, but without the harmonics.  Each output is a free-running LFO outputting one of 5 wave shapes (sine, square, triangle, saw, ramp) with a related-but-different clock speed; cv2 is slightly faster than cv1, cv3 is slightly faster than cv2, etc..., resulting in interesting phase shifting & drifting.

Inspired by Otterly by Expert Sleepers.

### CODE REVIEW

**Comment by chrisib** [comment]
I've made a small tweak to how the script works. Specifically the spread control now smoothly adjusts the speed of each channel relative to `cv1`:

- cv1 goes from x1 to x1 of cv1 (surprise! it's the master channel, and always goes at the same speed as itself!)
- cv2 will go from x1 to x1.2 of cv1
- cv3 will go from x1 to x1.25 of cv1
- cv4 will go from x1 to x1.333... of cv1
- cv5 will go from x1 to x1.5 of cv1
- cv6 will go from x1 to x2 of cv1

So at maximum spread each channel from cv2-6 is a nice harmonic interval of the speed of cv1: 1:1, 6:5, 5:4, 4:3, 3:2, and 2:1.

This fix makes the spread more predicable when changing the speed of cv1 via `k1` and/or `ain`.

File: software/contrib/lutra.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        self.display_pixels = [
+            [] for cv in cvs
+        ]
+
+        self.pixel_lock = _thread.allocate_lock()
+
+        self.load()
+
+    def load(self):
+        """Load and apply the saved configuration
```
Comment (mjaskula) [inline software/contrib/lutra.py:196]: [required] The configuration file and the saved state file are two different things in the europi firmware. It's confusing to have the documentation here call it a configuration file and then read from the saved state file. Ditto for the `save()` method below.

**Review by mjaskula — COMMENTED** [review]
Overall this looks pretty good to me. I like the addition of the multithreading documentation you put some good work into it.

I still want to actually run this script and play with it, and dig into the threading a little more, but here is an initial review.

File: software/contrib/lutra.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Lutra
+
+Lutra is a re-imagining of [Expert Sleepers' Otterly](https://expert-sleepers.co.uk/otterley.html) module.
+
+Each output channel is a free running LFO with a subtly different clock speed. The spread of clock speeds is controlled
+by `k2` (and optionally `ain` -- see configuration below)
+
+## Configuration
+
+The functionality of `ain` can be configured by creating/editing `/saved_state_Lutra.txt` on the module. This JSON file
```
[the remark this answers is not in the document]
Comment (chrisib) [reply software/contrib/lutra.md:10]: I interpreted things differently; I see the global config files (experimental_config & europi_config) as module-wide settings, while the saved_state files are effectively config files with a per-script focus.  Adding a script-specific setting to e.g. experimental_config feels like overkill, since that setting would never be used anywhere else.

Creating separate saved-state & config files for each script also feels counter-intuitive to me. The saved state _is_ the current configuration of the script. This particular feature is just a sneaky way of adding an extra bit of functionality that otherwise wouldn't be supported.

File: software/contrib/lutra.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+Turning `k2` fully anticlockwise will set the spread to zero; every output will have the same clock speed, (though
+depending on previous settings and random noise) they may be phase-shifted from each other.
+
+Turning `k2`clockwise will gradually increase the spread of clock speeds. `cv1` will always stay locked to the base
+clock speed, with `cv2-6` becoming progressively faster.
+
+## Re-syncing
+
+When `b1` or `din` receive a high voltage all CV outputs are temporarily halted.  Once the input drops low again
```
[the remark this answers is not in the document]
Comment (chrisib) [reply software/contrib/lutra.md:57]: I'll reword this to make it less developery-sounding.

File: software/firmware/experimental/thread.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        self.b1_falling = False
+        self.b2_rising = False
+        self.b2_falling = False
+
+        # High/low flags for each input
+        self.din_high = False
+        self.b1_high = False
+        self.b2_high = False
+
+        # MS ticks when events were last recorded
```
[the remark this answers is not in the document]
Comment (chrisib) [reply software/firmware/experimental/thread.py:58]: Yes, they're part of the API. They aren't used here but they are used in other multi-threaded scripts I have on the go. I'll add some additional documentation.

File: software/contrib/lutra.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+## Knob Control
+
+Turning `k1` fully anticlockwise will set the clock speed to the slowest setting. Turning `k1` fully clockwise will set
+the clock speed to the fastest setting.
+
+Turning `k2` fully anticlockwise will set the spread to zero; every output will have the same clock speed, (though
+depending on previous settings and random noise) they may be phase-shifted from each other.
+
+Turning `k2`clockwise will gradually increase the spread of clock speeds. `cv1` will always stay locked to the base
+clock speed, with `cv2-6` becoming progressively faster.  At maximum spread, the outpts follow common harmonic intervals
```
[the remark this answers is not in the document]
Comment (chrisib) [reply software/contrib/lutra.md:43]: An interesting though. Probably wouldn't be too hard to change the knob behaviour to allow the spread to speed up or slow down `cv2-6`.

File: software/contrib/lutra.py (hunk 1 above)
Comment (chrisib) [reply software/contrib/lutra.py:196]: Like I said above, the saved_state file (in my interpretation) _is_ the same thing as a per-script configuration file.

File: software/contrib/lutra.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjaskula) [reply software/contrib/lutra.md:10]: If that's how you are interpreting these features then the documentation for them needs to be improved. The firmware level config files are certainly module wide settings, but individual scripts can also have configuration, see the turing-machine and diagnostic script for examples.

> This particular feature is just a sneaky way of adding an extra bit of functionality that otherwise wouldn't be supported.

My assumptions about why this feature wouldn't otherwise be supported is that the limited number of hardware UI elements on the EuroPi make adding these features difficult with out making it very menu driven. This is explicitly one of the things that the configuration system was meant to address.

Please see the following issue and PRs for details of the motivations for these features. I will open a PR that pulls these details into the documentation.

https://github.com/Allen-Synthesis/EuroPi/issues/163
https://github.com/Allen-Synthesis/EuroPi/pull/220
https://github.com/Allen-Synthesis/EuroPi/pull/185

[the remark this answers is not in the document]
Comment (chrisib) [reply software/contrib/lutra.md:10]: Fair enough. I'll move that one item out of the saved state & create a separate config file for it.

File: software/contrib/lutra.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisib) [reply software/contrib/lutra.md]: Done.

File: software/contrib/lutra.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisib) [reply software/contrib/lutra.md]: I played around with this a little bit. In the end, I don't think I actually like this change; with the current implementation it's very easy to "lock" the waves' relative phases by just turning the spread control all the way to the left. That feels like a useful feature that gets lost if you need to centre the knob to set the spread to zero; the slightest imprecision will cause the waves to start spreading over time.

I may try it out some more later, but for now I think I'm leaning towards keeping the behaviour as it is right now.

File: software/contrib/lutra.py (hunk 1 above)
Comment (chrisib) [reply software/contrib/lutra.py]: Fixed

File: software/firmware/experimental/thread.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisib) [reply software/firmware/experimental/thread.py:75]: Added an additional section in the class header to explain the additional fields.

**Review by mjaskula — APPROVED** [review]
I never ended up spending anymore time with this, but that’s not a good reason to hold it up.
